crates/ra_hir/src/ty/infer.rs

```diff
@@ -421,6 +451,13 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 let typable: Option<TypableDef> = def.into();
                 let typable = typable?;
 
+                if let ExprOrPatId::Expr(expr) = id {
```

Comment:
hm, maybe it'd be better to do this above (after line 445), since
 - we only need to record resolutions for associated items
 - but we want to record resolutions for all kinds of associated items, and up there we still have the `ImplItem` (ok, the match needs to be changed to first return the `ImplItem`)
 - we should record resolutions for all intermediate associated items as well